File: backend/app/routers/usage.py
from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from ..db import DEFAULT_WORKSPACE_ID
from ..services.usage_tracker import (
    delete_pricing_rule,
    get_usage_summary,
    list_pricing_rules,
    list_usage_filters,
    save_pricing_rule,
)
from ._context import resolve_workspace_id

router = APIRouter()
ALL_WORKSPACES_SCOPE = "__all__"


def _resolve_usage_workspace_id(workspace_id: str | None) -> str | None:
    cleaned = str(workspace_id or "").strip()
    if not cleaned or cleaned == ALL_WORKSPACES_SCOPE:
        return None
    return resolve_workspace_id(cleaned)


class PricingRuleIn(BaseModel):
    provider: str
    model: str | None = None
    api_key_fingerprint: str | None = None
    input_price_per_1m: float = 0
    output_price_per_1m: float = 0
    currency: str = "USD"


@router.get("/filters")
def usage_filters(workspace_id: str = DEFAULT_WORKSPACE_ID) -> dict[str, list[str]]:
    return list_usage_filters(_resolve_usage_workspace_id(workspace_id))


@router.get("/summary")
def usage_summary(
    workspace_id: str = DEFAULT_WORKSPACE_ID,
    period: Literal["day", "month"] = Query(default="day"),
    breakdown_by: Literal["provider", "model", "feature", "api_key_fingerprint"] = Query(default="feature"),
    provider: str | None = None,
    model: str | None = None,
    feature: str | None = None,
    api_key: str | None = None,
) -> dict:
    return get_usage_summary(
        workspace_id=_resolve_usage_workspace_id(workspace_id),
        period=period,
        breakdown_by=breakdown_by,
        provider=provider,
        model=model,
        feature=feature,
        api_key_fingerprint_value=api_key,
    )


@router.get("/pricing")
def usage_pricing() -> list[dict]:
    return list_pricing_rules()


@router.put("/pricing")
def upsert_usage_pricing(payload: PricingRuleIn) -> dict:
    try:
        return save_pricing_rule(
            provider=payload.provider,
            model=payload.model,
            api_key_fingerprint_value=payload.api_key_fingerprint,
            input_price_per_1m=payload.input_price_per_1m,
            output_price_per_1m=payload.output_price_per_1m,
            currency=payload.currency,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.delete("/pricing/{rule_id}")
def remove_usage_pricing(rule_id: str) -> dict[str, bool]:
    return {"ok": delete_pricing_rule(rule_id)}

File: frontend-v3/src/shared/api/usage.ts
import { fetchJson } from "./http";

export const ALL_WORKSPACES_USAGE_SCOPE = "__all__";
export type UsagePeriod = "day" | "month";
export type UsageFeature = "indexing" | "translation" | "chat";
export type UsageBreakdownBy = "provider" | "model" | "feature" | "api_key_fingerprint";

export interface UsageFilters {
  providers: string[];
  models: string[];
  features: string[];
  api_keys: string[];
}

export interface UsageBucket {
  bucket: string;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  request_count: number;
  estimated_count: number;
  estimated_cost: number;
  dimension_breakdown: Record<string, number>;
}

export interface UsageTotals {
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  request_count: number;
  estimated_cost: number;
  dimension_breakdown: Record<string, number>;
  dimension_metrics: Record<string, {
    input_tokens: number;
    output_tokens: number;
    total_tokens: number;
    request_count: number;
    estimated_cost: number;
  }>;
}

export interface UsageSummary {
  period: UsagePeriod;
  breakdown_by: UsageBreakdownBy;
  available_range: {
    first_bucket: string;
    last_bucket: string;
  };
  buckets: UsageBucket[];
  totals: UsageTotals;
}

export interface PricingRule {
  id: string;
  provider: string;
  model?: string | null;
  api_key_fingerprint?: string | null;
  input_price_per_1m: number;
  output_price_per_1m: number;
  currency: string;
  created_at: string;
  updated_at: string;
}

export interface PricingRulePayload {
  provider: string;
  model?: string | null;
  api_key_fingerprint?: string | null;
  input_price_per_1m: number;
  output_price_per_1m: number;
  currency: string;
}

export async function listUsageFilters(workspaceId: string): Promise<UsageFilters> {
  const params = new URLSearchParams({ workspace_id: workspaceId });
  return fetchJson<UsageFilters>(`/api/usage/filters?${params.toString()}`);
}

export async function getUsageSummary(params: {
  workspaceId: string;
  period: UsagePeriod;
  breakdownBy: UsageBreakdownBy;
  startBucket?: string;
  endBucket?: string;
  provider?: string;
  model?: string;
  feature?: string;
  apiKey?: string;
}): Promise<UsageSummary> {
  const query = new URLSearchParams({ workspace_id: params.workspaceId, period: params.period, breakdown_by: params.breakdownBy });
  if (params.startBucket) query.set("start_bucket", params.startBucket);
  if (params.endBucket) query.set("end_bucket", params.endBucket);
  if (params.provider) query.set("provider", params.provider);
  if (params.model) query.set("model", params.model);
  if (params.feature) query.set("feature", params.feature);
  if (params.apiKey) query.set("api_key", params.apiKey);
  return fetchJson<UsageSummary>(`/api/usage/summary?${query.toString()}`);
}

export async function listPricingRules(): Promise<PricingRule[]> {
  return fetchJson<PricingRule[]>("/api/usage/pricing");
}

export async function savePricingRule(payload: PricingRulePayload): Promise<PricingRule> {
  return fetchJson<PricingRule>("/api/usage/pricing", {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export async function deletePricingRule(ruleId: string): Promise<{ ok: boolean }> {
  return fetchJson<{ ok: boolean }>(`/api/usage/pricing/${encodeURIComponent(ruleId)}`, { method: "DELETE" });
}
